pkgs/development/compilers/ldc/default.nix

```diff
@@ -18,208 +18,163 @@ let
+      # being undefined.
+      # TODO Remove with version > 0.17.6
+      substituteInPlace dmd2/root/port.c --replace __inline_isnanl __inline_isnan
+  '';
 
-      enableParallelBuilding = ldcBuild.enableParallelBuilding;
-      preferLocalBuild = true;
-      inputString = ldcBuild.outPath;
-      outputHashAlgo = "sha256";
-      outputHash = builtins.hashString "sha256" inputString;
+  nativeBuildInputs = [ cmake makeWrapper llvm unzip ]
 
-      src = ldcBuild.src;
+  ++ stdenv.lib.optional (!bootstrapVersion) [
+    bootstrapLdc python lit
+  ]
 
-      postUnpack = ldcBuild.postUnpack;
+  ++ stdenv.lib.optional (!bootstrapVersion && !stdenv.hostPlatform.isDarwin) [
+    # https://github.com/NixOS/nixpkgs/pull/36378#issuecomment-385034818
+    gdb
+  ]
 
-      postPatch = ldcBuild.postPatch;
+  ++ stdenv.lib.optional (bootstrapVersion) [
+    libconfig
+  ]
 
-      nativeBuildInputs = ldcBuild.nativeBuildInputs
+  ++ stdenv.lib.optional stdenv.hostPlatform.isDarwin (with darwin.apple_sdk.frameworks; [
+    Foundation
+  ]);
 
-      ++ [
-        ldcBuild
-      ];
 
-      buildInputs = ldcBuild.buildInputs;
+  buildInputs = [ curl tzdata ];
 
-      preConfigure = ''
-        cmakeFlagsArray=(
-          ${ldcBuild.cmakeFlagsString}
-          "-DD_COMPILER=${ldcBuild.out}/bin/ldmd2"
-        )
-      '';
+  cmakeFlagsString = stdenv.lib.optionalString (!bootstrapVersion) ''
+    "-DD_FLAGS=-d-version=TZDatabaseDir;-d-version=LibcurlPath;-J$PWD"
+    "-DCMAKE_BUILD_TYPE=Release"
+  '';
 
-      postConfigure = ldcBuild.postConfigure;
+  preConfigure = stdenv.lib.optionalString (!bootstrapVersion) ''
+    cmakeFlagsArray=(
+      ${cmakeFlagsString}
+    )
+  '';
 
-      makeFlags = ldcBuild.makeFlags;
+  postConfigure = ''
+    export DMD=$PWD/bin/ldmd2
+  '';
 
-      buildCmd = if bootstrapVersion then
-        "ctest -V -R \"build-druntime-ldc-unittest|build-phobos2-ldc-unittest\""
-      else
-        "make -j$NIX_BUILD_CORES DMD=${ldcBuild.out}/bin/ldc2 phobos2-test-runner phobos2-test-runner-debug";
+  makeFlags = [ "DMD=$DMD" ];
 
-      testCmd = if bootstrapVersion then
-        "ctest -j$NIX_BUILD_CORES --output-on-failure -E \"dmd-testsuite|lit-tests|ldc2-unittest|llvm-ir-testsuite\""
-      else
-        "ctest -j$NIX_BUILD_CORES --output-on-failure -E \"dmd-testsuite|lit-tests|ldc2-unittest\"";
+  fixNames = if stdenv.hostPlatform.isDarwin then ''
+    fixDarwinDylibNames() {
+      local flags=()
 
-      buildPhase = ''
-          ${buildCmd}
-          ln -s ${ldcBuild.out}/bin/ldmd2 $PWD/bin/ldmd2
-          ${testCmd}
-      '';
+      for fn in "$@"; do
+        flags+=(-change "$(basename "$fn")" "$fn")
+      done
 
-      installPhase = ''
-          echo -n $inputString > $out
-      '';
+      for fn in "$@"; do
+        if [ -L "$fn" ]; then continue; fi
+        echo "$fn: fixing dylib"
+        install_name_tool -id "$fn" "''${flags[@]}" "$fn"
+      done
     }
```

Comment:
I see this is a direct copypasta form `pkgs/build-support/setup-hooks/fix-darwin-dylib-names.sh`.

I think what you want can be achieved by moving your test to `installCheckPhase`. If you look at the default order phases are executed
https://github.com/NixOS/nixpkgs/blob/358efb66947504d2195a4f1e25e72265cc7348e3/pkgs/stdenv/generic/setup.sh#L1264-L1267
`installCheckPhase` goes after `fixupPhase`. You can let `fixupPhase` do the `install_name_tool` for you before your test is run in `installCheckPhase`.

Reply:
I tried that, it doesn't work as I need the source files for the unit tests and those are not available in this stage anymore.

Reply:
Okay.